File: cpp/Maximum_Side_Length_of_a_Square_with_Sum_Less_than_or_Equal_to_Threshold_1292.cpp
class Solution {
public:
    int maxSideLength(vector<vector<int>>& mat, int threshold) {
        
        for (int i = 1; i < mat.size(); i++) {
            mat[i][0] += mat[i - 1][0];
        }
        for (int i = 1; i < mat[0].size(); i++) {
            mat[0][i] += mat[0][i - 1];
        }
        for (int i = 1; i < mat.size(); i++) {
            for (int j = 1; j < mat[i].size(); j++) {
                mat[i][j] += mat[i - 1][j] + mat[i][j - 1] - mat[i - 1][j - 1];
            }
        }

        auto calcMinArea = [&mat](int side) {
            int minArea = static_cast<int>(1e9);
            for (int i = side - 1; i < mat.size(); i++) {
                for (int j = side - 1; j < mat[i].size(); j++) {
                int curArea = mat[i][j] - (i >= side ? mat[i - side][j] : 0) - (j >= side ? mat[i][j - side] : 0) + (i >= side && j >= side ? mat[i - side][j - side] : 0);
                    minArea = min(minArea, curArea);
                }
            }
            return minArea;
        };

        int left = 0, right = min(mat.size(), mat[0].size()), mid;
        while (true) {
            mid = (left + right) / 2;
            if (right - left < 2) {
                if (calcMinArea(right) <= threshold) return right;
                return left;
            }
            if (calcMinArea(mid) > threshold) right = mid;
            else left = mid;
        }

        return -1;

    }
};

File: python/Maximum_Side_Length_of_a_Square_with_Sum_Less_than_or_Equal_to_Threshold_1292.py
class Solution:
    def maxSideLength(self, mat: List[List[int]], threshold: int) -> int:

        for i in range(len(mat)):
            for j in range(len(mat[0])):
                mat[i][j] += (mat[i - 1][j] if i > 0 else 0) + (mat[i][j - 1] if j > 0 else 0) - (mat[i - 1][j - 1] if i > 0 and j > 0 else 0)
        
        def calcMinArea(side):
            minArea = 1e9
            for i in range(side - 1, len(mat)):
                for j in range(side - 1, len(mat[0])):
                    curArea = mat[i][j] - (mat[i - side][j] if i >= side else 0) - (mat[i][j - side] if j >= side else 0) + (mat[i - side][j - side] if i >= side and j >= side else 0)
                    minArea = min(minArea, curArea)
            return minArea
        
        left, right, mid = 0, min(len(mat), len(mat[0])), 0
        while True:
            mid = (left + right) // 2
            if right - left < 2:
                if calcMinArea(right) <= threshold:
                    return right
                return left
            if calcMinArea(mid) > threshold:
                right = mid
            else:
                left = mid
        
        return -1
